### PR TITLE
[1LP][RFR] Automate tests for C&U data rollup in Availability zones

### PR DESCRIPTION
**Pull request by nachandr**
## Purpose or Intent
Added tests for  C&U data rollup in Availability zones. The tests check if the hourly rollup records are available in the DB.

### PRT Run
{{py.test: cfme/tests/candu/test_utilization_metrics.py --use-provider complete -k "azone"}}

### PRT Results:
510 - PASS
511 - PASS

### CODE REVIEW

**Comment by dajoRH** [comment]
Lint report for commit c2e3d549841e29ae845aad4ab64d82899ffeb012:

`cfme/tests/candu/test_utilization_metrics.py`:
- :bangbang: Line 324:9: **E117** *over-indented*
- :bangbang: Line 328:9: **E112** *expected an indented block*
- :bangbang: Line 328:10: **E999** *IndentationError: expected an indented block*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
I detected some fixture changes in commit 643d8e2631298a721dc70f974af1ac4e14567687

The *local* fixture **`metrics_collection`** is used in the following files:
- **cfme/tests/candu/test_utilization_metrics.py**
    - *test_raw_metric_vm_cpu*
    - *test_raw_metric_vm_memory*
    - *test_raw_metric_vm_network*
    - *test_raw_metric_vm_disk*
    - *test_raw_metric_host_cpu*
    - *test_raw_metric_host_memory*
    - *test_raw_metric_host_network*
    - *test_raw_metric_host_disk*
    - *test_azone_candu_cpu*
    - *test_azone_candu_memory*
    - *test_azone_candu_network*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

File: cfme/tests/candu/test_utilization_metrics.py [hunk 1]
```diff
@@ -302,3 +308,71 @@ def test_raw_metric_host_disk(metrics_collection, appliance, provider):
...
+def test_azone_candu_network(metrics_collection, appliance, provider):
+    """
+    Polarion:
+        assignee: nachandr
+        caseimportance: high
+        casecomponent: CandU
+        initialEstimate: 1/12h
+    """
+    azone_name = provider.data["cap_and_util"]["capandu_azone"]
+    generic_test_azone_rollup(appliance, provider, 'net_usage_rate_average', azone_name)
```
Comment (john-dupuy) [inline cfme/tests/candu/test_utilization_metrics.py]: @nachandr **Required** since these tests use the same providers and steps I recommend parametrizing them like

```python
@pytest.mark.parametrize("metric", ['cpu_usage_rate_average', 'mem_usage_absolute_average', 'net_usage_rate_average'], ids=['cpu', 'mem', 'network'])
```

File: cfme/tests/candu/test_utilization_metrics.py [hunk 2]
```diff
@@ -302,3 +308,71 @@ def test_raw_metric_host_disk(metrics_collection, appliance, provider):
...
+        )
+    return appliance.db.client.session.query(metrics_tbl).filter(
+        metrics_tbl.id.in_(provs.subquery()))
+
+
+def generic_test_azone_rollup(appliance, provider, metric, azone_name):
+    query = query_metric_rollup_table(appliance, provider, metric, azone_name)
+
+    for record in query:
+        if getattr(record, metric) is not None:
```
Comment (john-dupuy) [inline cfme/tests/candu/test_utilization_metrics.py]:
```suggestion
        if hasattr(record, metric):
```

File: cfme/tests/candu/test_utilization_metrics.py [hunk 3]
```diff
@@ -302,3 +308,71 @@ def test_raw_metric_host_disk(metrics_collection, appliance, provider):
...
+            assert getattr(record, metric) > 0, (
```
Comment (john-dupuy) [inline cfme/tests/candu/test_utilization_metrics.py]: Since we now know that `record` has the attr of `metric`

```suggestion
            assert record.metric > 0, (
```

File: cfme/tests/candu/test_utilization_metrics.py [hunk 4]
```diff
@@ -129,6 +134,7 @@ def test_raw_metric_vm_cpu(metrics_collection, appliance, provider):
             vm_name)
         average_rate = attrgetter('cpu_usage_rate_average')
 
+    pytest.set_trace()
```
Comment (john-dupuy) [inline cfme/tests/candu/test_utilization_metrics.py]: Is this for debugging purposes? Still necessary?

File: cfme/tests/candu/test_utilization_metrics.py (hunk 1 above)
Comment (nachandr) [reply cfme/tests/candu/test_utilization_metrics.py]: The three new tests require that the metrics_collection fixture be run. The first metric collection is scheduled through this  fixture. The first metric collection itself takes at least 15 minutes(sometimes longer). The three tests that are run per provider take at least  20 minutes each for completion(considering  provider add, provider refresh etc.). But, without parametrization, the three tests could bee run in under 20 minutes.So, I did not parametrize the tests.

I should have added a note about this in the PR when I submitted the PR.

File: cfme/tests/candu/test_utilization_metrics.py (hunk 4 above)
Comment (nachandr) [reply cfme/tests/candu/test_utilization_metrics.py]: Removed the trace statement.

**Comment by dajoRH** [comment]
I detected some fixture changes in commit 4bfda299d6b61a31794deb3c15d6520bef009ba8

The *local* fixture **`metrics_collection`** is used in the following files:
- **cfme/tests/candu/test_utilization_metrics.py**
    - *test_raw_metric_vm_cpu*
    - *test_raw_metric_vm_memory*
    - *test_raw_metric_vm_network*
    - *test_raw_metric_vm_disk*
    - *test_raw_metric_host_cpu*
    - *test_raw_metric_host_memory*
    - *test_raw_metric_host_network*
    - *test_raw_metric_host_disk*
    - *test_azone_candu_cpu*
    - *test_azone_candu_memory*
    - *test_azone_candu_network*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

File: cfme/tests/candu/test_utilization_metrics.py (hunk 1 above)
Comment (john-dupuy) [reply cfme/tests/candu/test_utilization_metrics.py]: Ah okay, thanks for the explanation!

File: cfme/tests/candu/test_utilization_metrics.py [hunk 5]
```diff
@@ -302,3 +316,69 @@ def test_raw_metric_host_disk(metrics_collection, appliance, provider):
...
+    for record in query:
+        if hasattr(record, metric):
+            assert record.metric > 0, (
+                'Zero Azone CPU usage/memory usage/Disk IO/Network IO'
+            )
+            break
+
+
+@pytest.mark.uncollectif(lambda provider:
+                         provider.one_of(RHEVMProvider, VMwareProvider, OpenStackProvider))
```
Comment (john-dupuy) [inline cfme/tests/candu/test_utilization_metrics.py]: These and the subsequent provider filters can be moved to the top of the file. Also regarding the `uncollectif`, I expect @mshriver will comment something similar here https://github.com/ManageIQ/integration_tests/pull/9506#discussion_r339072512

Comment (nachandr) [reply cfme/tests/candu/test_utilization_metrics.py]: pytest.mark.provider( [EC2Provider, AzureProvider], required_fields=[(['cap_and_util', 'capandu_vm'], 'cu-24x7')], 

> scope="module")``

Comment (nachandr) [reply cfme/tests/candu/test_utilization_metrics.py]: I have replaced the uncollectif statement in the tests.

**Comment by dajoRH** [comment]
I detected some fixture changes in commit 477ba8423fa4270b4f035fe47d88dde3a4e073c7

The *local* fixture **`metrics_collection`** is used in the following files:
- **cfme/tests/candu/test_utilization_metrics.py**
    - *test_raw_metric_vm_cpu*
    - *test_raw_metric_vm_memory*
    - *test_raw_metric_vm_network*
    - *test_raw_metric_vm_disk*
    - *test_raw_metric_host_cpu*
    - *test_raw_metric_host_memory*
    - *test_raw_metric_host_network*
    - *test_raw_metric_host_disk*
    - *test_azone_candu_cpu*
    - *test_azone_candu_memory*
    - *test_azone_candu_network*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

**Comment by dajoRH** [comment]
I detected some fixture changes in commit 266d7af13ebf08db015a6fc7f39729e266130e19

The *local* fixture **`metrics_collection`** is used in the following files:
- **cfme/tests/candu/test_utilization_metrics.py**
    - *test_raw_metric_vm_cpu*
    - *test_raw_metric_vm_memory*
    - *test_raw_metric_vm_network*
    - *test_raw_metric_vm_disk*
    - *test_raw_metric_host_cpu*
    - *test_raw_metric_host_memory*
    - *test_raw_metric_host_network*
    - *test_raw_metric_host_disk*
    - *test_azone_candu_cpu*
    - *test_azone_candu_memory*
    - *test_azone_candu_network*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

**Comment by dajoRH** [comment]
I detected some fixture changes in commit 773e55d4678970b4689ec73a44417a9084d71639

The *local* fixture **`metrics_collection`** is used in the following files:
- **cfme/tests/candu/test_utilization_metrics.py**
    - *test_raw_metric_vm_cpu*
    - *test_raw_metric_vm_memory*
    - *test_raw_metric_vm_network*
    - *test_raw_metric_vm_disk*
    - *test_raw_metric_host_cpu*
    - *test_raw_metric_host_memory*
    - *test_raw_metric_host_network*
    - *test_raw_metric_host_disk*
    - *test_azone_cpu_usage*
    - *test_azone_memory_usage*
    - *test_azone_network_io*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

File: cfme/tests/candu/test_utilization_metrics.py [hunk 6]
```diff
@@ -290,7 +304,7 @@ def test_raw_metric_host_disk(metrics_collection, appliance, provider):
     """
     Polarion:
         assignee: nachandr
-        caseimportance: medium
+        caseimportance: high
```
Comment (mshriver) [inline cfme/tests/candu/test_utilization_metrics.py]: :+1:

File: cfme/tests/candu/test_utilization_metrics.py [hunk 7]
```diff
@@ -302,3 +316,90 @@ def test_raw_metric_host_disk(metrics_collection, appliance, provider):
         if record.disk_usage_rate_average is not None:
             assert record.disk_usage_rate_average > 0, 'Zero Host Disk IO'
             break
+
+
+def query_metric_rollup_table(appliance, provider, metric, azone_name):
+    metrics_tbl = appliance.db.client['metric_rollups']
```
Comment (mshriver) [inline cfme/tests/candu/test_utilization_metrics.py]: This needs a docblock explaining in plain English what the intent of the query is, and what's being returned.

File: cfme/tests/candu/test_utilization_metrics.py (hunk 3 above)
Comment (mshriver) [reply cfme/tests/candu/test_utilization_metrics.py]: Can we take this one step further and use getattr's default?

Secondly, concern about the TypeError that will result here from the attribute value not being an int?  Easy to wrap it and avoid that just because there's a string.

```
assert int(getattr(record, metric, 0)) > 0
```

File: cfme/tests/candu/test_utilization_metrics.py [hunk 8]
```diff
@@ -302,3 +316,90 @@ def test_raw_metric_host_disk(metrics_collection, appliance, provider):
...
+            .join(ems, metrics_tbl.parent_ems_id == ems.id)
+            .filter(metrics_tbl.resource_name == azone_name,
+            ems.name == provider.name)
+        )
+    return appliance.db.client.session.query(metrics_tbl).filter(
+        metrics_tbl.id.in_(provs.subquery()))
+
+
+def generic_test_azone_rollup(appliance, provider, metric):
+    azone_name = provider.data["cap_and_util"]["capandu_azone"]
```
Comment (mshriver) [inline cfme/tests/candu/test_utilization_metrics.py]: Again, docblock please of the intent.

This looks to be asserting against the first record in the query that includes the metric, and returning nothing?

File: cfme/tests/candu/test_utilization_metrics.py [hunk 9]
```diff
@@ -302,3 +316,90 @@ def test_raw_metric_host_disk(metrics_collection, appliance, provider):
...
+@pytest.mark.provider(
+    [EC2Provider, AzureProvider],
+    required_fields=[(['cap_and_util', 'capandu_vm'], 'cu-24x7')],
+    override=True, scope="module"
+)
+@pytest.mark.meta(
+    blockers=[BZ(1744845, forced_streams=['5.10'],
+        unblock=lambda provider: not provider.one_of(AzureProvider))]
+)
+def test_azone_cpu_usage(metrics_collection, appliance, provider):
```
Comment (mshriver) [inline cfme/tests/candu/test_utilization_metrics.py]: Consider parametrizing here, as it looks like all three test functions have the same provider marks, same BZ blocking, same polarion meta, and same function call, just with a different string for the rollup metric.

Could just parametrize those metric names and have a single function, so all these provider marks aren't repeated so much.

That or wrap them in a class at least so you can apply the provider mark to the class.

Comment (john-dupuy) [reply cfme/tests/candu/test_utilization_metrics.py]: @mshriver https://github.com/ManageIQ/integration_tests/pull/9530#discussion_r339027325

**Comment by dajoRH** [comment]
I detected some fixture changes in commit b5b4e9e1254fe22f5f68a52391a21495708da480

The *local* fixture **`metrics_collection`** is used in the following files:
- **cfme/tests/candu/test_utilization_metrics.py**
    - *test_raw_metric_vm_cpu*
    - *test_raw_metric_vm_memory*
    - *test_raw_metric_vm_network*
    - *test_raw_metric_vm_disk*
    - *test_raw_metric_host_cpu*
    - *test_raw_metric_host_memory*
    - *test_raw_metric_host_network*
    - *test_raw_metric_host_disk*
    - *test_azone_cpu_usage*
    - *test_azone_memory_usage*
    - *test_azone_network_io*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

File: cfme/tests/candu/test_utilization_metrics.py [hunk 10]
```diff
@@ -36,6 +37,7 @@ def clean_setup_provider(request, provider):
 
 
 def vm_count(appliance, metrics_tbl, mgmt_system_id):
+    pytest.set_trace()
```
Comment (john-dupuy) [inline cfme/tests/candu/test_utilization_metrics.py]: Leftover from some debugging? Can this be removed?

File: cfme/tests/candu/test_utilization_metrics.py (hunk 3 above)
Comment (nachandr) [reply cfme/tests/candu/test_utilization_metrics.py]: Done

File: cfme/tests/candu/test_utilization_metrics.py (hunk 7 above)
Comment (nachandr) [reply cfme/tests/candu/test_utilization_metrics.py]: Done

File: cfme/tests/candu/test_utilization_metrics.py (hunk 8 above)
Comment (nachandr) [reply cfme/tests/candu/test_utilization_metrics.py]: Done

File: cfme/tests/candu/test_utilization_metrics.py (hunk 10 above)
Comment (nachandr) [reply cfme/tests/candu/test_utilization_metrics.py]: Yes, of course, that needs to go :)

File: cfme/tests/candu/test_utilization_metrics.py [hunk 11]
```diff
@@ -302,3 +316,86 @@ def test_raw_metric_host_disk(metrics_collection, appliance, provider):
...
+    haven't been parametrized.
+
+    In an effort to reduce DRY code, I've written this generic function that
+    checks if any of the table records have a metric with a non-zero value.
+    """
+    azone_name = provider.data["cap_and_util"]["capandu_azone"]
+    query = query_metric_rollup_table(appliance, provider, metric, azone_name)
+
+    for record in query:
+        if hasattr(record, metric) and int(getattr(record, metric, 0)):
```
Comment (mshriver) [inline cfme/tests/candu/test_utilization_metrics.py]: can drop the hasattr, getattr() with default won't raise.

File: cfme/tests/candu/test_utilization_metrics.py [hunk 12]
```diff
@@ -302,3 +316,86 @@ def test_raw_metric_host_disk(metrics_collection, appliance, provider):
...
+            return True
+        else:
+            raise ValueError('The record had a zero in it!')
+
+
+@pytest.mark.provider(
+    [EC2Provider, AzureProvider],
+    required_fields=[(['cap_and_util', 'capandu_vm'], 'cu-24x7')],
+    override=True, scope="module"
```
Comment (mshriver) [inline cfme/tests/candu/test_utilization_metrics.py]: override is no longer needed here - pytest5 markers take care of it for us.

**Comment by dajoRH** [comment]
I detected some fixture changes in commit 2f355433a33869d9a9b8fe23e21584c634f27ce5

The *local* fixture **`metrics_collection`** is used in the following files:
- **cfme/tests/candu/test_utilization_metrics.py**
    - *test_raw_metric_vm_cpu*
    - *test_raw_metric_vm_memory*
    - *test_raw_metric_vm_network*
    - *test_raw_metric_vm_disk*
    - *test_raw_metric_host_cpu*
    - *test_raw_metric_host_memory*
    - *test_raw_metric_host_network*
    - *test_raw_metric_host_disk*
    - *test_azone_cpu_usage*
    - *test_azone_memory_usage*
    - *test_azone_network_io*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

**Comment by dajoRH** [comment]
I detected some fixture changes in commit 6091a49f0dc74576408423b06cb388e2d42710fd

The *local* fixture **`metrics_collection`** is used in the following files:
- **cfme/tests/candu/test_utilization_metrics.py**
    - *test_raw_metric_vm_cpu*
    - *test_raw_metric_vm_memory*
    - *test_raw_metric_vm_network*
    - *test_raw_metric_vm_disk*
    - *test_raw_metric_host_cpu*
    - *test_raw_metric_host_memory*
    - *test_raw_metric_host_network*
    - *test_raw_metric_host_disk*
    - *test_azone_cpu_usage*
    - *test_azone_memory_usage*
    - *test_azone_network_io*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:
